x/logic/meter/weighted_meter.go

```diff
@@ -0,0 +1,90 @@
+package meter
+
+import (
+	"math"
+
+	"github.com/cosmos/cosmos-sdk/types"
+)
+
+// WeightedMeterDecorator is decorator that wraps a gas meter and adds a weight multiplier to the consumed gas.
+type WeightedMeterDecorator struct {
+	decorated types.GasMeter
+	weight    uint64
+}
+
+// WithWeightedMeter returns a new WeightedMeterDecorator with the given weight.
+func WithWeightedMeter(decorated types.GasMeter, weight uint64) types.GasMeter {
+	return &WeightedMeterDecorator{
+		decorated: decorated,
+		weight:    weight,
+	}
+}
+
+// GasConsumed returns the amount of gas consumed by the decorated gas meter.
+func (m *WeightedMeterDecorator) GasConsumed() types.Gas {
+	return m.decorated.GasConsumed()
+}
+
+// GasConsumedToLimit returns the amount of gas consumed by the decorated gas meter.
+func (m *WeightedMeterDecorator) GasConsumedToLimit() types.Gas {
+	return m.decorated.GasConsumedToLimit()
+}
+
+// GasRemaining returns the amount of gas remaining in the decorated gas meter.
+func (m *WeightedMeterDecorator) GasRemaining() types.Gas {
+	return m.decorated.GasRemaining()
+}
+
+// Limit returns the limit of the decorated gas meter.
+func (m *WeightedMeterDecorator) Limit() types.Gas {
+	return m.decorated.Limit()
+}
+
+// ConsumeGas consumes the given amount of gas from the decorated gas meter.
+func (m *WeightedMeterDecorator) ConsumeGas(amount types.Gas, descriptor string) {
+	consumed, overflow := multiplyUint64Overflow(m.decorated.GasConsumed(), amount)
```

Comment:
I think the amount should be multiplied by `m.weight` instead.

Reply:
That's totally right. My bad. The fact is the meter is not put under test. My bad x2.